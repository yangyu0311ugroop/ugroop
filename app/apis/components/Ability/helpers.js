import { LOGIC_HELPERS } from 'utils/helpers/logic';
import { TOUR_CONTRIBUTOR_ROLE } from './roles';

const logicalUserRole = (allowedRoles, tourRole, role) => {
  if (!allowedRoles.includes(role)) {
    return LOGIC_HELPERS.ifElse(
      tourRole === TOUR_CONTRIBUTOR_ROLE.TOUR_ORGANIZER,
      tourRole,
      '',
    );
  }

  return role;
};

const enableCheckbox = (id, ownerCanAccess, canCreate, isAccessible) =>
  LOGIC_HELPERS.ifElse(
    [id && ownerCanAccess, canCreate, isAccessible],
    true,
    false,
    true,
  );

const showNote = (id, ownerCanAccess, canCreate, isAccessible) => {
  const checkboxEnabled = enableCheckbox(
    id,
    ownerCanAccess,
    canCreate,
    isAccessible,
  );

  return LOGIC_HELPERS.ifElse([id, !checkboxEnabled], true, false);
};

const getRoleUsingRelatedIds = (relatedIds, id) => {
  let role = '';
  const filterRelatedIds = relatedIds.filter(array => array[2] === id);
  if (filterRelatedIds.length === 0) {
    return role;
  }

  const filterRole = filterRelatedIds.filter(
    array => array[1] === 'tour_organizer',
  );
  role = filterRole.length === 0 ? role : filterRole[0][1];
  return role;
};

const showCheckbox = (canCreate, hasOrgAccess, isTabCreator) =>
  canCreate || hasOrgAccess || isTabCreator;

export const ABILITY_HELPERS = {
  logicalUserRole,
  enableCheckbox,
  showNote,
  showCheckbox,
  getRoleUsingRelatedIds,
};
