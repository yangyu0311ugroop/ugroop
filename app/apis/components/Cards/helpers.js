import {
  CARDS_API,
  GET_CUSTOMER_CARDS,
  CREATE_CUSTOMER_CARD,
  UPDATE_CUSTOMER_CARD,
  DELETE_CUSTOMER_CARD,
} from 'apis/constants';
import { get } from 'lodash';
import { normalize, schema } from 'normalizr';

const cards = new schema.Entity('cards');

const customers = new schema.Entity('customers', {
  cards: [cards],
});

const createCustomerCard = (obj, props) => {
  const { id, data, onSuccess, onError } = obj;
  props.resaga.dispatchTo(CARDS_API, CREATE_CUSTOMER_CARD, {
    payload: { id, data },
    onSuccess,
    onError,
  });
};

const deleteCustomerCard = (obj, props) => {
  const { id, cardId, isDefault, onSuccess, onError } = obj;
  props.resaga.dispatchTo(CARDS_API, DELETE_CUSTOMER_CARD, {
    payload: { id, cardId, isDefault },
    onSuccess,
    onError,
  });
};

const updateCustomerCard = (obj, props) => {
  const { id, cardId, data, onSuccess, onError } = obj;
  props.resaga.dispatchTo(CARDS_API, UPDATE_CUSTOMER_CARD, {
    payload: { id, cardId, data },
    onSuccess,
    onError,
  });
};

const getCustomerCards = (obj, props) => {
  const { id, onSuccess, onError } = obj;
  props.resaga.dispatchTo(CARDS_API, GET_CUSTOMER_CARDS, {
    payload: { id },
    onSuccess,
    onError,
  });
};

const normaliseDeleteCard = (data, payload) => ({
  payload,
  raw: data,
});

const normaliseCreateCard = (data, payload) => ({
  raw: data,
  payload,
});
const normaliseUpdateCard = (data, payload) => ({
  raw: data,
  payload,
});

const normaliseGetCard = (result, payload) => {
  const { id } = payload;
  const customerData = {
    id,
    cards: get(result, 'data', []),
  };
  const { entities } = normalize(customerData, CARDS_SCHEMA.customers);

  return {
    resultCards: get(entities.customers, `${id}.cards`),
    normalised: {
      customers: entities.customers,
      cards: entities.cards,
    },
    payload,
  };
};

export const CARD_API_HELPERS = {
  createCustomerCard,
  updateCustomerCard,
  deleteCustomerCard,
  getCustomerCards,
};

export const CARD_NORMALISER = {
  normaliseCreateCard,
  normaliseDeleteCard,
  normaliseUpdateCard,
  normaliseGetCard,
};
export const CARDS_SCHEMA = {
  customers,
};
