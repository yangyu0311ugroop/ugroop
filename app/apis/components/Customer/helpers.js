import { CUSTOMER_API, UPDATE_CUSTOMER } from 'apis/constants';

const updateCustomer = (obj, props) => {
  const { id, data, onSuccess, onError } = obj;
  props.resaga.dispatchTo(CUSTOMER_API, UPDATE_CUSTOMER, {
    payload: { id, data },
    onSuccess,
    onError,
  });
};

export const CUSTOMER_API_HELPERS = {
  updateCustomer,
};
