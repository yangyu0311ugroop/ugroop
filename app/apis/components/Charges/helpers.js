import { STRIPE_FETCH_THRESHOLD } from 'appConstants';
import { CHARGES_API, GET_CUSTOMER_CHARGES } from 'apis/constants';
import { DATASTORE_UTILS } from 'datastore';
import { get, last, set } from 'lodash';
import { normalize } from 'normalizr';
import { CHARGES_SCHEMA } from 'datastore/charges/schema';
import { LOGIC_HELPERS } from 'utils/helpers/logic';

const getCustomerCharges = (obj, props) => {
  const { data, onSuccess, onError } = obj;
  props.resaga.dispatchTo(CHARGES_API, GET_CUSTOMER_CHARGES, {
    payload: {
      data: { ...data, limit: STRIPE_FETCH_THRESHOLD.CHARGES_GET_LIMIT },
    },
    onSuccess,
    onError,
  });
};

const handleListCharges = (
  entityCustomer,
  customerData,
  payload,
) => storeData => {
  const id = get(payload, 'data.customer');
  const chargePath = `${id}.charges`;
  const newCharges = get(entityCustomer, chargePath);

  if (!newCharges && !id) return storeData;
  const newStore = DATASTORE_UTILS.upsertArray(chargePath, newCharges)(
    storeData,
  );
  set(newStore, `${id}.listChargesHasMore`, customerData.listChargesHasMore);
  set(newStore, `${id}.lastFetchId`, customerData.lastFetchId);
  return newStore;
};

const normaliseListCharges = (data, payload) => {
  const charges = get(data, 'charges.data');
  const hasMore = get(data, 'charges.has_more', false);
  const lastFetchCharges = LOGIC_HELPERS.ifElse(
    charges && charges.length > 0,
    last(charges),
  );
  const lastFetchId = get(lastFetchCharges, 'id');
  const customerData = {
    id: get(payload, 'data.customer'),
    listChargesHasMore: hasMore,
    lastFetchId,
    charges,
  };
  const { entities } = normalize(customerData, CHARGES_SCHEMA.customers);
  return {
    normalised: {
      customers: entities.customers,
      charges: entities.charges,
    },
    raw: data,
    payload,
  };
};

export const CHARGES_API_HELPERS = {
  getCustomerCharges,
  handleListCharges,
};

export const CHARGES_NORMALISER = {
  normaliseListCharges,
};
