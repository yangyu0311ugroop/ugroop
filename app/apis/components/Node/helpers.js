import {
  NODE_API,
  GET_NODE,
  CREATE_CHILD,
  CREATE_NEXT_NODE,
  UPDATE_NODE,
  DELETE_CHILDREN,
  GET_TIMES,
  GET_TREE,
  INSERT_AFTER,
  INSERT_BEFORE,
  DELETE_TEMP_NODE,
  SHARE_NODE,
  CREATE_ATTACHMENT,
  UPDATE_ATTACHMENT,
  REMOVE_ATTACHMENT,
  MOVE_AFTER,
  UNLINK_NEXT_NODE,
} from 'apis/constants';
import { DO_NOTHING, EMPTY_RTE } from 'appConstants';
import { ARRAY_HELPERS } from 'utils/helpers/arrays';
import { CONSOLE_HELPERS, MISSING_REQUIRED_PROPS } from 'utils/helpers/console';
import { DATASTORE_UTILS } from 'datastore';
import { TOUR_ROLES } from 'datastore/invitationStore/constants';
import { NODE_STORE_HELPERS } from 'datastore/nodeStore/helpers';
import { LOGIC_HELPERS } from 'utils/helpers/logic';
import { requests } from 'utils/request';

/**
 * Gets type-checked node based on id.
 * Use this if you need customData.
 * @param obj
 *  id: node id
 *  type: node type
 *  childKey: specify to automatically update parent node's id array
 *  ...rest (eg. onSuccess, onError, etc)
 * @param props
 *  resaga: any resaga instance
 */
const getNode = (obj, props) => {
  const { id, type, updateParent = true, childKey, ...rest } = obj;

  props.resaga.dispatchTo(NODE_API, GET_NODE, {
    payload: {
      id,
      type,
      childKey: LOGIC_HELPERS.ifElse(updateParent, childKey || 'children'),
    },
    ...rest,
  });
};

// call this function to either create child or next node
const createNode = (
  {
    node,
    parentNodeId,
    childKey = 'children',
    onSuccess,
    onError,
    lastNodeId,
    fetchCalculated,
    insertLocation = '',
  },
  ownProps,
) => {
  // required props
  if (!parentNodeId) {
    return CONSOLE_HELPERS.error(MISSING_REQUIRED_PROPS, 'parentNodeId');
  }

  const { resaga } = ownProps;

  const handleSuccess = fetchCalculated
    ? ({ node: responseNode }) => {
        NODE_API_HELPERS.getTimes(
          {
            id: responseNode.id,
            ids: [responseNode.id],
            onSuccess,
            onError,
          },
          ownProps,
        );
      }
    : onSuccess;

  if (!lastNodeId) {
    return resaga.dispatchTo(NODE_API, CREATE_CHILD, {
      payload: {
        keyPath: `${parentNodeId}.${childKey}`,
        nodeId: parentNodeId,
        parentNodeId,
        node,
      },
      onSuccess: handleSuccess,
      onError,
    });
  }

  if (insertLocation === 'after') {
    return resaga.dispatchTo(NODE_API, INSERT_AFTER, {
      payload: {
        id: lastNodeId,
        parentId: parentNodeId,
        insertLocation: 'after',
        node,
        type: childKey,
      },
      onSuccess: handleSuccess,
      onError,
    });
  }

  if (insertLocation === 'before') {
    return resaga.dispatchTo(NODE_API, INSERT_BEFORE, {
      payload: {
        id: lastNodeId,
        parentId: parentNodeId,
        insertLocation: 'before',
        node,
        type: childKey,
      },
      onSuccess: handleSuccess,
      onError,
    });
  }

  return resaga.dispatchTo(NODE_API, CREATE_NEXT_NODE, {
    payload: {
      keyPath: `${parentNodeId}.${childKey}`,
      nodeId: lastNodeId,
      parentNodeId,
      node,
    },
    onSuccess: handleSuccess,
    onError,
  });
};

const updateNode = ({ nodeId, node, onSuccess, onError }, props) => {
  const { resaga } = props;
  return resaga.dispatchTo(NODE_API, UPDATE_NODE, {
    payload: { nodeId, node },
    onSuccess,
    onError,
  });
};

const deleteNode = (
  {
    nodeId,
    dependentIds,
    dependentLinkIds,
    childKey = 'children',
    parent,
    type,
    onSuccess,
    onError,
  },
  props,
) => {
  const { parentNodeId: parentProp, resaga } = props;
  const parentNodeId = parent || parentProp;

  // required props
  if (!nodeId) {
    return CONSOLE_HELPERS.error(MISSING_REQUIRED_PROPS, 'nodeId');
  }
  if (!parentNodeId) {
    return CONSOLE_HELPERS.error(MISSING_REQUIRED_PROPS, 'parentNodeId');
  }

  return resaga.dispatchTo(NODE_API, DELETE_CHILDREN, {
    payload: {
      keyPath: `${parentNodeId}.${childKey}`,
      nodeId,
      type,
      dependentIds,
      dependentLinkIds,
    },
    onSuccess,
    onError,
  });
};

const deleteTempNode = (
  { nodeId, childKey = 'children', parent, onSuccess, onError },
  props,
) => {
  const { parentNodeId: parentProp, resaga } = props;
  const parentNodeId = parent || parentProp;

  // required props
  if (!nodeId) {
    return CONSOLE_HELPERS.error(MISSING_REQUIRED_PROPS, 'nodeId');
  }
  if (!parentNodeId) {
    return CONSOLE_HELPERS.error(MISSING_REQUIRED_PROPS, 'parentNodeId');
  }

  return resaga.dispatchTo(NODE_API, DELETE_TEMP_NODE, {
    payload: {
      keyPath: `${parentNodeId}.${childKey}`,
      nodeId,
    },
    onSuccess,
    onError,
  });
};

const convertNextNode = ({ nextNodeId, childKey = 'children' }, props) => {
  if (!nextNodeId) {
    return DO_NOTHING;
  }

  const { resaga } = props;

  return resaga.setValue({
    [childKey]: NODE_STORE_HELPERS.concatIfNotExist(nextNodeId), // add nextNodeId to childKey
    nextNodes: NODE_STORE_HELPERS.deleteIfExist(nextNodeId), // delete nextNodeId from nextNodes
  });
};

/**
 * @param obj
 *  id: node id for permission check
 *  ids: one or more node ids for which to get times
 *  onSuccess: success callback
 *  onError: error callback
 * @param props
 *  resaga: any resaga instance
 */
const getTimes = (obj, props) => {
  const { id, ids, onSuccess, onError } = obj;
  return props.resaga.dispatchTo(NODE_API, GET_TIMES, {
    payload: { id, ids },
    onSuccess,
    onError,
  });
};

/**
 * @param obj
 *  id: node id
 *  idUpsertMode: array upsert method (append by default)
 *  onSuccess: success callback
 *  onError: error callback
 * @param props
 *  resaga: any resaga instance
 */
const getTree = (obj, props) => {
  const { id, idUpsertMode, onSuccess, onError } = obj;
  props.resaga.dispatchTo(NODE_API, GET_TREE, {
    payload: { id, idUpsertMode },
    onSuccess,
    onError,
  });
};

/**
 * @param obj
 *  id: node id
 *  idUpsertMode: array upsert method (append by default)
 *  onSuccess: success callback
 *  onError: error callback
 * @param props
 *  resaga: any resaga instance
 */
const getTreeAndTimes = (obj, props) => {
  const { id, idUpsertMode, onSuccess, onError } = obj;

  if (!id) return null;

  // TODO: Combine in one api call?
  return props.resaga.dispatchTo(NODE_API, GET_TREE, {
    payload: { id, idUpsertMode },
    onSuccess: ({ node, eventNodes }) => {
      const ids = [
        ...DATASTORE_UTILS.getObjectIds(node),
        ...DATASTORE_UTILS.getObjectIds(eventNodes),
      ];
      NODE_API_HELPERS.getTimes(
        {
          id: ids[0],
          ids,
          onSuccess,
          onError,
        },
        props,
      );
    },
    onError,
  });
};

/**
 *
 * @param obj
 *  id: node id
 *  role: new invitee role
 *  shareTo: invitee address
 *  shareToUserId: invitee user id (optional)
 *  content: message to invitee (optional)
 *  inviteToOrganisation: whether to also invite to node's org (optional)
 *  subNodes: extra {nodeId, role}'s to include in invitation (optional)
 *  ...rest (eg. onSuccess, onError, etc)
 * @param props
 *  resaga: any resaga instance
 */
const shareNode = (obj, props) => {
  const {
    id,
    role,
    shareTo,
    shareToUserId,
    content,
    inviteToOrganisation = false,
    subNodes,
    fullName,
    ...rest
  } = obj;

  const payload = {
    id,
    payload: {
      role,
      content: content === EMPTY_RTE ? '' : content,
      shareTo,
      fullName,
      roleName: TOUR_ROLES[role],
      inviteToOrganisation,
      subNodes: ARRAY_HELPERS.arrayIfNot(subNodes),
    },
    shareToUserId,
  };

  props.resaga.dispatchTo(NODE_API, SHARE_NODE, { payload, ...rest });
};

/**
 * @param obj
 *  id: node id
 *  file: { name, url, size }
 *  ...rest (eg. onSuccess, onError, etc)
 * @param props
 *  resaga: any resaga instance
 */
const createAttachment = (obj, props) => {
  const {
    id,
    file: { name, url, size, fileSize, description },
    ...rest
  } = obj;

  props.resaga.dispatchTo(NODE_API, CREATE_ATTACHMENT, {
    payload: {
      id,
      name,
      fileSize: size || fileSize,
      description,
      url,
      isSection: true, // TODO: Properly rename isSection to isNode?
    },
    ...rest,
  });
};

/**
 * @param obj
 *  id: node id,
 *  attachmentId,
 *  name: name of file,
 *  url: path to file,
 *  fileSize: size of file,
 *  description: description of file,
 *  ...rest (eg. onSuccess, onError, etc)
 * @param props
 *  resaga: any resaga instance
 */
const updateAttachment = (obj, props) => {
  const { id, attachmentId, name, url, fileSize, description, ...rest } = obj;
  props.resaga.dispatchTo(NODE_API, UPDATE_ATTACHMENT, {
    payload: {
      id,
      attachmentId,
      name,
      url,
      fileSize,
      description,
      isSection: true, // TODO: Properly rename isSection to isNode?
    },
    ...rest,
  });
};

/**
 * @param obj
 *  id: node id,
 *  ...rest (eg. onSuccess, onError, etc)
 * @param props
 *  resaga: any resaga instance
 */
const removeAttachment = (obj, props) => {
  const { id, ...rest } = obj;
  props.resaga.dispatchTo(NODE_API, REMOVE_ATTACHMENT, {
    payload: { id, isSection: true }, // TODO: Properly rename isSection to isNode?
    ...rest,
  });
};

/**
 * @param obj
 *  node: { type, content, ... }
 *  parentNodeId
 *  childKey
 *  file: { name, url, size }
 *  ...rest (eg. onSuccess, onError, etc)
 * @param props
 *  resaga: any resaga instance
 */
const createNodeAndAttachment = (obj, props) => {
  // TODO: Combine in one api call?
  NODE_API_HELPERS.createNode(
    {
      ...obj,
      onSuccess: ({ node: { id } }) =>
        NODE_API_HELPERS.createAttachment({ ...obj, id }, props),
    },
    props,
  );
};

/**
 * @param obj
 *  nodeId: node id
 *  node: { type, content, ... }
 *  attachmentId
 *  attachment: { name, url, fileSize, description }
 *  ...rest (eg. onSuccess, onError, etc)
 * @param props
 *  resaga: any resaga instance
 */
const updateNodeAndAttachment = (
  { nodeId, node, attachmentId, attachment, ...rest },
  props,
) => {
  // TODO: Combine in one api call?
  NODE_API_HELPERS.updateNode(
    {
      nodeId,
      node,
      onSuccess: () =>
        NODE_API_HELPERS.updateAttachment(
          {
            ...attachment,
            attachmentId,
            id: nodeId,
            ...rest,
          },
          props,
        ),
    },
    props,
  );
};

const moveNodeAfter = (
  { nodeId, nodeIdToBeMoved, removeParentNode = true, onSuccess, onError },
  props,
) => {
  props.resaga.dispatchTo(NODE_API, MOVE_AFTER, {
    payload: {
      id: nodeId,
      toBeMovedId: nodeIdToBeMoved,
      removeParentNode,
    },
    onSuccess,
    onError,
  });
};

const unlinkNextNode = (
  { nodeId, nodeToBeUnlinkedId, path, onSuccess, onError },
  props,
) => {
  props.resaga.dispatchTo(NODE_API, UNLINK_NEXT_NODE, {
    payload: {
      id: nodeId,
      fk: nodeToBeUnlinkedId,
      path,
    },
    onSuccess,
    onError,
  });
};

const dispatchUpdateNode = node =>
  requests.fetchWithAuthorisation('patch', `/${NODE_API}/${node.id}`, node);

const batchUpdateNode = ({ nodeId, node }) => {
  if (Array.isArray(node)) {
    return Promise.all(node.map(NODE_API_HELPERS.dispatchUpdateNode));
  }

  return NODE_API_HELPERS.dispatchUpdateNode({ ...node, id: nodeId });
};

export const NODE_API_HELPERS = {
  getNode,
  createNode,
  updateNode,
  batchUpdateNode,
  deleteNode,
  deleteTempNode,
  convertNextNode,
  getTimes,
  getTree,
  getTreeAndTimes,
  shareNode,
  createAttachment,
  updateAttachment,
  removeAttachment,
  createNodeAndAttachment,
  updateNodeAndAttachment,
  moveNodeAfter,
  unlinkNextNode,
  dispatchUpdateNode,
};
