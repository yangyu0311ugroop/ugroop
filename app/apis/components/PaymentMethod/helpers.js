import { PAYMENT_METHOD_API, UPDATE_PAYMENT_METHOD } from 'apis/constants';
import { DATASTORE_UTILS } from 'datastore';
import { LOGIC_HELPERS } from 'utils/helpers/logic';

const updatePaymentMethod = (obj, props) => {
  const { id, data, onSuccess, onError } = obj;
  props.resaga.dispatchTo(PAYMENT_METHOD_API, UPDATE_PAYMENT_METHOD, {
    payload: { id, data },
    onSuccess,
    onError,
  });
};

const normaliseUpdate = ({ paymentMethod: data }, payload) => {
  const { billing_details: billingAddress, card, ...rest } = data;

  const { address } = billingAddress;

  const addressReduced = Object.keys(address).reduce(
    (acc, key) => ({
      ...acc,
      ...{
        [`address_${LOGIC_HELPERS.ifElse(
          key === 'postal_code',
          'zip',
          key,
        )}`]: address[key],
      },
    }),
    {},
  );

  return {
    paymentSources: DATASTORE_UTILS.upsertObject({
      [rest.id]: { ...card, ...rest, ...addressReduced },
    }),
    raw: data,
    payload,
    card: { ...card },
    rest: { ...rest },
    address: { ...addressReduced },
  };
};

export const PAYMENT_METHOD_API_HELPERS = {
  updatePaymentMethod,
};

export const PAYMENT_METHOD_NORMALISER = {
  normaliseUpdate,
};
