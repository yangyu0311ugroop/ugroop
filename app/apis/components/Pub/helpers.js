import {
  GET_PUB_TEMPLATE_TIMES,
  GET_PUB_TEMPLATE_TREE,
  GET_PUB_TEMPLATE_EVENTS,
  PUB_CREATE_INTEREST,
  PUB_API,
} from 'apis/constants';
import { DATASTORE_UTILS } from 'datastore';
import { RESAGA_HELPERS } from 'utils/helpers/resaga';
import { NODE_STORE_SELECTORS } from 'datastore/nodeStore/selectors';

/**
 * @param obj
 *  hashkey: template hashkey
 *  onSuccess: success callback
 *  onError: error callback
 * @param props
 *  resaga: any resaga instance
 */
const fetchEvents = (obj, props) => {
  const { hashkey, onSuccess, onError } = obj;
  props.resaga.dispatchTo(PUB_API, GET_PUB_TEMPLATE_EVENTS, {
    payload: {
      hashkey,
    },
    onSuccess,
    onError,
  });
};

/**
 * @param obj
 *  hashkey: template hashkey
 *  idUpsertMode: id array upsert method (append by default)
 *  onSuccess: success callback
 *  onError: error callback
 * @param props
 *  resaga: any resaga instance
 */
const getTreeAndTimes = (obj, props) => {
  const { hashkey, idUpsertMode, onSuccess, onError } = obj;
  // TODO: Combine in one api call?
  props.resaga.dispatchTo(PUB_API, GET_PUB_TEMPLATE_TREE, {
    payload: { hashkey, idUpsertMode },
    onSuccess: ({ node, eventNodes }) => {
      const ids = [
        ...DATASTORE_UTILS.getObjectIds(node),
        ...DATASTORE_UTILS.getObjectIds(eventNodes),
      ];
      props.resaga.dispatchTo(PUB_API, GET_PUB_TEMPLATE_TIMES, {
        payload: { hashkey, ids },
        onSuccess,
        onError,
      });
    },
    onError,
  });
};

/**
 * @param obj
 *  hashkey: template hashkey
 *  data: node + paticipant
 *  onSuccess: success callback
 *  onError: error callback
 * @param props
 *  resaga: any resaga instance
 */
const createInterest = (obj, props) => {
  const { hashkey, data, onSuccess, onError } = obj;
  props.resaga.dispatchTo(PUB_API, PUB_CREATE_INTEREST, {
    payload: {
      hashkey,
      data,
    },
    onSuccess,
    onError,
  });
};

export const CONFIG = {
  value: {
    hashkeyDescription: RESAGA_HELPERS.mapToId(
      NODE_STORE_SELECTORS.hashkeyDescription,
      'templateId',
    ),
    orgId: RESAGA_HELPERS.mapToId(
      NODE_STORE_SELECTORS.organisationId,
      'templateId',
    ),
  },
};

export const PUB_API_HELPERS = {
  fetchEvents,
  getTreeAndTimes,
  createInterest,
};
