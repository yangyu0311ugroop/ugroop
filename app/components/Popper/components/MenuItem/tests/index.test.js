import { shallow } from 'enzyme';
import React from 'react';
import { LOGIC_HELPERS } from 'utils/helpers/logic';
import { TEST_HELPERS } from 'utils/helpers/testHelpers';
import { MenuItem } from '../index';

describe('<MenuItem />', () => {
  let rendered;
  let instance;

  const resaga = {
    setValue: jest.fn(),
    dispatchTo: jest.fn(),
  };

  const props = {
    classes: { default: 'default' },
    resaga,
  };

  beforeEach(() => {
    rendered = shallow(<MenuItem {...props} />);
    instance = rendered.instance();
  });

  it('should exists', () => {
    expect(MenuItem).toBeDefined();
  });

  it('should render without exploding', () => {
    expect(rendered.length).toBe(1);
  });

  describe('handleClickMenu()', () => {
    it('should handleClickMenu()', () => {
      LOGIC_HELPERS.ifFunction = jest.fn();
      const stopPropagation = jest.fn();

      instance.handleClickMenu({ stopPropagation });

      TEST_HELPERS.expectCalledAndMatchSnapshot(LOGIC_HELPERS.ifFunction);
      TEST_HELPERS.expectCalledAndMatchSnapshot(stopPropagation);
    });

    it('should not stopPropagation', () => {
      LOGIC_HELPERS.ifFunction = jest.fn();
      const stopPropagation = jest.fn();
      rendered.setProps({ stopPropagation: false });

      instance.handleClickMenu({ stopPropagation });

      TEST_HELPERS.expectCalledAndMatchSnapshot(LOGIC_HELPERS.ifFunction);
      expect(stopPropagation).not.toBeCalled();
    });
  });

  describe('renderIcon()', () => {
    it('should return null', () => {
      rendered.setProps({ icon: '' });

      expect(instance.renderIcon()).toBe(null);
    });

    it('should renderIcon', () => {
      rendered.setProps({ icon: 'icon' });

      TEST_HELPERS.expectMatchSnapshot(instance.renderIcon);
    });
  });

  describe('renderMenuItem()', () => {
    it('should renderMenuItem', () => {
      rendered.setProps({ compact: true });
      instance.renderIcon = jest.fn(() => 'renderIcon');

      TEST_HELPERS.expectMatchSnapshot(instance.renderMenuItem);
    });
  });

  describe('renderDefault()', () => {
    it('should render div', () => {
      rendered.setProps({ onClick: null });

      instance.renderMenuItem = jest.fn(() => 'renderMenuItem');

      TEST_HELPERS.expectMatchSnapshot(instance.renderDefault);
    });

    it('should render button', () => {
      rendered.setProps({ onClick: jest.fn() });

      instance.renderMenuItem = jest.fn(() => 'renderMenuItem');

      TEST_HELPERS.expectMatchSnapshot(instance.renderDefault);
    });
  });

  describe('render()', () => {
    it('should return LOGIC_HELPERS.switchCase', () => {
      LOGIC_HELPERS.switchCase = jest.fn(() => 'switchCase');

      expect(instance.render()).toBe('switchCase');

      TEST_HELPERS.expectCalledAndMatchSnapshot(LOGIC_HELPERS.switchCase);
    });
  });
});
