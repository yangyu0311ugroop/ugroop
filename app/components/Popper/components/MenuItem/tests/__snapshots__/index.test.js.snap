// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MenuItem /> handleClickMenu() should handleClickMenu() 1`] = `
Array [
  Array [
    undefined,
    Array [
      Object {
        "stopPropagation": [MockFunction] {
          "calls": Array [
            Array [],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        },
      },
      Object {
        "IconProps": Object {},
        "alignItems": "center",
        "block": true,
        "className": "",
        "classes": Object {
          "default": "default",
        },
        "color": "default",
        "compact": false,
        "disabled": false,
        "icon": null,
        "resaga": Object {
          "dispatchTo": [MockFunction],
          "setValue": [MockFunction],
        },
        "type": null,
        "variant": "",
      },
    ],
  ],
  Array [
    undefined,
    Array [
      Object {
        "stopPropagation": [MockFunction] {
          "calls": Array [
            Array [],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        },
      },
      Object {
        "IconProps": Object {},
        "alignItems": "center",
        "block": true,
        "className": "",
        "classes": Object {
          "default": "default",
        },
        "color": "default",
        "compact": false,
        "disabled": false,
        "icon": null,
        "resaga": Object {
          "dispatchTo": [MockFunction],
          "setValue": [MockFunction],
        },
        "type": null,
        "variant": "",
      },
    ],
  ],
]
`;

exports[`<MenuItem /> handleClickMenu() should handleClickMenu() 2`] = `
Array [
  Array [],
]
`;

exports[`<MenuItem /> handleClickMenu() should not stopPropagation 1`] = `
Array [
  Array [
    undefined,
    Array [
      Object {
        "stopPropagation": [MockFunction],
      },
      Object {
        "IconProps": Object {},
        "alignItems": "center",
        "block": true,
        "className": "",
        "classes": Object {
          "default": "default",
        },
        "color": "default",
        "compact": false,
        "disabled": false,
        "icon": null,
        "resaga": Object {
          "dispatchTo": [MockFunction],
          "setValue": [MockFunction],
        },
        "type": null,
        "variant": "",
      },
    ],
  ],
  Array [
    undefined,
    Array [
      Object {
        "stopPropagation": [MockFunction],
      },
      Object {
        "IconProps": Object {},
        "alignItems": "center",
        "block": true,
        "className": "",
        "classes": Object {
          "default": "default",
        },
        "color": "default",
        "compact": false,
        "disabled": false,
        "icon": null,
        "resaga": Object {
          "dispatchTo": [MockFunction],
          "setValue": [MockFunction],
        },
        "type": null,
        "variant": "",
      },
    ],
  ],
]
`;

exports[`<MenuItem /> render() should return LOGIC_HELPERS.switchCase 1`] = `
Array [
  Array [
    "",
    Object {
      "default": [Function],
    },
  ],
]
`;

exports[`<MenuItem /> renderDefault() should render button 1`] = `<div />`;

exports[`<MenuItem /> renderDefault() should render div 1`] = `<div />`;

exports[`<MenuItem /> renderIcon() should renderIcon 1`] = `<div />`;

exports[`<MenuItem /> renderMenuItem() should renderMenuItem 1`] = `<div />`;
